Core/NoOpRunner.Core/Shapes/BaseShape.cs

```diff
@@ -74,5 +74,29 @@ public virtual void OnClick()
 
         public abstract bool CanOverlap(BaseShape other);
         public abstract void OnCollision(BaseShape other);
+
+        public BaseShape Clone()
+        {
+            return (BaseShape)this.MemberwiseClone();
+        }
+        public BaseShape DeepCopy()
+        {
+            BaseShape clone = (BaseShape)this.MemberwiseClone();
+            clone.ShapeBlocks = new List<ShapeBlock>();
+            foreach (ShapeBlock shape in ShapeBlocks)
+            {
+                ShapeBlock block = new ShapeBlock();
+                block.OffsetX = shape.OffsetX;
+                block.OffsetY = shape.OffsetY;
+                clone.ShapeBlocks.Add(block);
+            }
```

Comment:
Could just be `clone.ShapeBlocks = ShapeBlocks.Select(x => new ShapeBlock() { OffsetX = x.OffsetX, OffsetY = x.OffsetY }).ToList();`